hail/src/main/scala/is/hail/expr/ir/Simplify.scala

```diff
@@ -222,28 +222,50 @@ object Simplify {
     case GetTupleElement(MakeTuple(xs), idx) => xs(idx)
 
     case TableCount(TableMapGlobals(child, _)) => TableCount(child)
-
     case TableCount(TableMapRows(child, _)) => TableCount(child)
-
     case TableCount(TableRepartition(child, _, _)) => TableCount(child)
-
     case TableCount(TableUnion(children)) =>
       children.map(TableCount).reduce[IR](ApplyBinaryPrimOp(Add(), _, _))
-
     case TableCount(TableKeyBy(child, _, _)) => TableCount(child)
-
     case TableCount(TableOrderBy(child, _)) => TableCount(child)
-
     case TableCount(TableLeftJoinRightDistinct(child, _, _)) => TableCount(child)
-
     case TableCount(TableRange(n, _)) => I64(n)
-
     case TableCount(TableParallelize(rows, _)) => Cast(ArrayLen(rows), TInt64())
-
     case TableCount(TableRename(child, _, _)) => TableCount(child)
-
     case TableCount(TableAggregateByKey(child, _)) => TableCount(TableDistinct(child))
 
+    // TableGetGlobals should simplify very aggressively
+    case TableGetGlobals(child) if child.typ.globalType == TStruct() => MakeStruct(FastSeq())
+    case TableGetGlobals(TableKeyBy(child, _, _)) => TableGetGlobals(child)
+    case TableGetGlobals(TableFilter(child, _)) => TableGetGlobals(child)
+    case TableGetGlobals(TableHead(child, _)) => TableGetGlobals(child)
+    case TableGetGlobals(TableRepartition(child, _, _)) => TableGetGlobals(child)
+    case TableGetGlobals(TableJoin(child, _, _, _)) => TableGetGlobals(child)
```

Comment:
TableJoin actually concatenates the globals of its children.

Reply:
oof, will make sure this is tested